ibis/pyspark/compiler.py

```diff
@@ -1052,32 +1055,53 @@ def compile_window_op(t, expr, scope, timecontext, **kwargs):
     ]
 
     order_by = window._order_by
+    # Timestamp needs to be cast to long for filtering in spark
     ordering_keys = [
-        key.to_expr().get_name()
+        F.col(key.to_expr().get_name()).cast('long')
+        if isinstance(key.args[0], types.TimestampColumn)
+        else key.to_expr().get_name()
         for key in map(operator.methodcaller('op'), order_by)
     ]
-
     context = AggregationContext.WINDOW
     pyspark_window = Window.partitionBy(grouping_keys).orderBy(ordering_keys)
 
     # If the operand is a shift op (e.g. lead, lag), Spark will set the window
     # bounds. Only set window bounds here if not a shift operation.
     if not isinstance(operand.op(), ops.ShiftBase):
-        start = (
-            -window.preceding
-            if window.preceding is not None
-            else Window.unboundedPreceding
-        )
-        end = (
-            window.following
-            if window.following is not None
-            else Window.unboundedFollowing
-        )
-        pyspark_window = pyspark_window.rowsBetween(start, end)
+        if window.preceding is None:
+            start = Window.unboundedPreceding
+        elif isinstance(window.preceding, ir.IntervalScalar):
+            start = -int(execute(window.preceding).value / 1e9)
+        elif isinstance(window.preceding, int):
+            start = -window.preceding
+        else:
+            raise com.UnsupportedOperationError(
+                'preceding type {}is not supported in windowing'.format(
+                    type(window.preceding)
+                )
+            )
+        if window.following is None:
+            end = Window.unboundedFollowing
+        elif isinstance(window.following, ir.IntervalScalar):
+            end = int(execute(window.following).value / 1e9)
```

Comment:
These are duplicate logic as `window.preceding`. Can you refactor?

Reply:
Sure, I added a function to canonicalize these intervals.